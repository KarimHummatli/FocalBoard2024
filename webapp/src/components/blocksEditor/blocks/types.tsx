import {ContentBlockTypes} from '../../../blocks/block'

export type BlockInputProps = {
    onChange: (value: string) => void
    value: string
    onCancel: () => void
    onSave: (val: string) => void
}

export type ContentType = {
    name: ContentBlockTypes
    displayName: string
    slashCommand: string
    prefix: string
    Input: React.FunctionComponent<BlockInputProps>
    render: (value: string) => React.ReactNode
    runSlashCommand: (changeType: (contentType: ContentType) => void, changeValue: (value: string) => void, ...args: string[]) => void
    nextType?: string
}
