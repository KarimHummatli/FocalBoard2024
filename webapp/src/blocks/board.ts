// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.
import {Utils} from '../utils'

import {Block, IBlock} from './block'
import {Card} from './card'

type PropertyType = 'text' | 'number' | 'select' | 'multiSelect' | 'date' | 'person' | 'file' | 'checkbox' | 'url' | 'email' | 'phone' | 'createdTime' | 'createdBy' | 'updatedTime' | 'updatedBy'

interface IPropertyOption {
    id: string
    value: string
    color: string
}

// A template for card properties attached to a board
interface IPropertyTemplate {
    id: string
    name: string
    type: PropertyType
    options: IPropertyOption[]
}

type BoardFields = {
    icon: string
    description: string
    showDescription?: boolean
    isTemplate?: boolean
    cardProperties: IPropertyTemplate[]
}

class Board extends Block {
    fields: BoardFields

    constructor(block?: IBlock) {
        super(block)
        this.type = 'board'

        let cardProperties: IPropertyTemplate[] = []
        if (block?.fields.cardProperties) {
            // Deep clone of card properties and their options
            cardProperties = block?.fields.cardProperties.map((o: IPropertyTemplate) => {
                return {
                    id: o.id,
                    name: o.name,
                    type: o.type,
                    options: o.options ? o.options.map((option) => ({...option})) : [],
                }
            })
        }

        const selectProperties = cardProperties.find((o) => o.type === 'select')
        if (!selectProperties) {
            const property: IPropertyTemplate = {
                id: Utils.createGuid(),
                name: 'Status',
                type: 'select',
                options: [],
            }
            cardProperties.push(property)
        }

        this.fields = {
            showDescription: block?.fields.showDescription || false,
            description: block?.fields.description || '',
            icon: block?.fields.icon || '',
            isTemplate: block?.fields.isTemplate || false,
            cardProperties,
        }
    }

    duplicate(): Board {
        const board = new Board(this)
        board.id = Utils.createGuid()
        return board
    }
}

type BoardGroup = {
    option: IPropertyOption
    cards: Card[]
}

export {Board, PropertyType, IPropertyOption, IPropertyTemplate, BoardGroup}
