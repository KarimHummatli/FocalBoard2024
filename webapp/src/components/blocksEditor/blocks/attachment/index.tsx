import React, {useRef, useEffect} from 'react'
import {BlockInputProps, ContentType} from '../types'

import './attachment.scss'

const Attachment: ContentType = {
    name: 'attachment',
    displayName: 'Attachment',
    slashCommand: '/attachment',
    prefix: '',
    runSlashCommand: (): void => {},
    editable: false,
    Display: (props: BlockInputProps) => <div className='AttachmentView'>ðŸ“Ž {props.value}</div>,
    Input: (props: BlockInputProps) => {
        const ref = useRef<HTMLInputElement|null>(null)
        useEffect(() => {
            ref.current?.click()
        }, [])

        return (
            <input
                ref={ref}
                className='Attachment'
                type='file'
                onChange={(e) => {
                    const files = e.currentTarget?.files
                    if (files) {
                        for (let i = 0; i < files.length; i++) {
                            const file = files.item(i)
                            if (file) {
                                props.onSave(file.name as string)
                            }
                        }
                    }
                }}
            />
        )
    }
}

Attachment.runSlashCommand = (changeType: (contentType: ContentType) => void, changeValue: (value: string) => void): void => {
    changeType(Attachment)
    changeValue('')
}

export default Attachment
