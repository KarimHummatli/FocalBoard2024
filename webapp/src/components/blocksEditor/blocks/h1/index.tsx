import React, {useRef, useEffect} from 'react'
import {marked} from 'marked'

import {BlockInputProps, ContentType} from '../types'

import './h1.scss'

const H1: ContentType = {
    name: 'h1',
    displayName: 'Title',
    slashCommand: '/title',
    prefix: '# ',
    render: (value: string) => {
        const renderer = new marked.Renderer()
        const html = marked('# '+value, {renderer, breaks: true})
        return <div
            dangerouslySetInnerHTML={{__html: html.trim()}}
        />
    },
    runSlashCommand: (): void => {},
    Input: (props: BlockInputProps) => {
        const ref = useRef<HTMLInputElement|null>(null)
        useEffect(() => {
            ref.current?.focus()
        }, [])
        return (
            <input
                ref={ref}
                className='H1'
                onChange={(e) => props.onChange(e.currentTarget.value)}
                onKeyDown={(e) => {
                    if (props.value === '' && e.key === "Backspace") {
                        props.onCancel()
                    }
                    if (e.key === "Enter") {
                        props.onSave(props.value)
                    }
                }}
                value={props.value}
            />
        )
    }
}

H1.runSlashCommand = (changeType: (contentType: ContentType) => void, changeValue: (value: string) => void, ...args: string[]): void => {
    changeType(H1)
    changeValue(args.join(' '))
}

export default H1
