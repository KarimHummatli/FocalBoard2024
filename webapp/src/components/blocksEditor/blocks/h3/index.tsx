import React, {useRef, useEffect} from 'react'
import {marked} from 'marked'

import {BlockInputProps, ContentType} from '../types'

import './h3.scss'

const H3: ContentType = {
    name: 'h3',
    displayName: 'Sub Sub title',
    slashCommand: '/subsubtitle',
    prefix: '### ',
    render: (value: string) => {
        const renderer = new marked.Renderer()
        const html = marked('### '+value, {renderer, breaks: true})
        return <div
            dangerouslySetInnerHTML={{__html: html.trim()}}
        />
    },
    runSlashCommand: (): void => {},
    Input: (props: BlockInputProps) => {
        const ref = useRef<HTMLInputElement|null>(null)
        useEffect(() => {
            ref.current?.focus()
        }, [])
        return (
            <input
                ref={ref}
                className='H3'
                onChange={(e) => props.onChange(e.currentTarget.value)}
                onKeyDown={(e) => {
                    if (props.value === '' && e.key === "Backspace") {
                        props.onCancel()
                    }
                    if (e.key === "Enter") {
                        props.onSave(props.value)
                    }
                }}
                value={props.value}
            />
        )
    }
}

H3.runSlashCommand = (changeType: (contentType: ContentType) => void, changeValue: (value: string) => void, ...args: string[]): void => {
    changeType(H3)
    changeValue(args.join(' '))
}

export default H3
