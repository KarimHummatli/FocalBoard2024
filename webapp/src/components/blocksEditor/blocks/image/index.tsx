import React, {useRef, useEffect} from 'react'
import {BlockInputProps, ContentType} from '../types'

import './image.scss'

type FileInfo = {
    file: string|ArrayBuffer
    width?: number
    align?: 'left'|'center'|'right'
}

const Image: ContentType<FileInfo> = {
    name: 'image',
    displayName: 'Image',
    slashCommand: '/image',
    prefix: '',
    runSlashCommand: (): void => {},
    editable: false,
    Display: (props: BlockInputProps<FileInfo>) => {
        if (props.value.file) {
            return <img className='ImageView' src={props.value.file as string}/>
        }
        return null
    },
    Input: (props: BlockInputProps<FileInfo>) => {
        const ref = useRef<HTMLInputElement|null>(null)
        useEffect(() => {
            ref.current?.click()
        }, [])

        return (
            <div>
                {props.value.file && (typeof props.value.file === 'string') && (
                    <img
                        className='ImageView' src={props.value.file}
                        onClick={() => ref.current?.click()}
                    />
                )}
                <input
                    ref={ref}
                    className='Image'
                    type='file'
                    accept='image/*'
                    onChange={(e) => {
                        const reader = new FileReader();
                        const file = (e.currentTarget?.files || [])[0]
                        reader.readAsArrayBuffer(file);
                        reader.onload = function () {
                            props.onSave({file: reader.result as ArrayBuffer})
                        }
                    }}
                />
            </div>
        )
    }
}

Image.runSlashCommand = (changeType: (contentType: ContentType<FileInfo>) => void, changeValue: (value: FileInfo) => void): void => {
    changeType(Image)
    changeValue({file: ''})
}

export default Image
