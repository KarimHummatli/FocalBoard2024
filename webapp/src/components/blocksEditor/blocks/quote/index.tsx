import React, {useRef, useEffect} from 'react'
import {marked} from 'marked'

import {BlockInputProps, ContentType} from '../types'

import './quote.scss'

const Quote: ContentType = {
    name: 'quote',
    displayName: 'Quote',
    slashCommand: '/quote',
    prefix: '> ',
    render: (value: string) => {
        const renderer = new marked.Renderer()
        const html = marked('> '+value, {renderer, breaks: true})
        return <div
            className='Quote'
            dangerouslySetInnerHTML={{__html: html.trim()}}
        />
    },
    runSlashCommand: (): void => {},
    Input: (props: BlockInputProps) => {
        const ref = useRef<HTMLInputElement|null>(null)
        useEffect(() => {
            ref.current?.focus()
        }, [])
        return (
            <blockquote className='Quote'>
                <input
                    ref={ref}
                    onChange={(e) => props.onChange(e.currentTarget.value)}
                    onKeyDown={(e) => {
                        if (props.value === '' && e.key === "Backspace") {
                            props.onCancel()
                        }
                        if (e.key === "Enter") {
                            props.onSave(props.value)
                        }
                    }}
                    onBlur={() => props.onSave(props.value)}
                    value={props.value}
                />
            </blockquote>
        )
    }
}

Quote.runSlashCommand = (changeType: (contentType: ContentType) => void, changeValue: (value: string) => void, ...args: string[]): void => {
    changeType(Quote)
    changeValue(args.join(' '))
}

export default Quote
