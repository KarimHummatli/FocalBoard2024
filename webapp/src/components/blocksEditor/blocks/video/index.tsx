import React, {useRef, useEffect} from 'react'
import {BlockInputProps, ContentType} from '../types'

import './video.scss'

const Video: ContentType = {
    name: 'video',
    displayName: 'Video',
    slashCommand: '/video',
    prefix: '',
    render: (value: string) => (
        <video width="320" height="240" controls className='VideoView'>
            <source src={value} type='video/mp4'/>
        </video>
    ),
    runSlashCommand: (): void => {},
    Input: (props: BlockInputProps) => {
        const ref = useRef<HTMLInputElement|null>(null)
        useEffect(() => {
            ref.current?.click()
        }, [])

        return (
            <input
                ref={ref}
                className='Video'
                type='file'
                accept='video/*'
                onChange={(e) => {
                    const reader = new FileReader();
                    const file = (e.currentTarget?.files || [])[0]
                    reader.readAsDataURL(file);
                    reader.onload = function () {
                        props.onSave(reader.result as string)
                    }
                }}
            />
        )
    }
}

Video.runSlashCommand = (changeType: (contentType: ContentType) => void, changeValue: (value: string) => void): void => {
    changeType(Video)
    changeValue('')
}

export default Video
