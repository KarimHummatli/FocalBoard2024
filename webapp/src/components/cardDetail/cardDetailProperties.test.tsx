// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.
import {fireEvent, render} from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import 'isomorphic-fetch'

import React from 'react'

import {IPropertyTemplate} from '../../blocks/board'
import {FetchMock} from '../../test/fetchMock'
import {TestBlockFactory} from '../../test/testBlockFactory'

import {wrapIntl} from '../../testUtils'

import CardDetailProperties from './cardDetailProperties'

global.fetch = FetchMock.fn

beforeEach(() => {
    FetchMock.fn.mockReset()
})

describe('components/cardDetail/CardDetailProperties', () => {
    const board = TestBlockFactory.createBoard()
    board.fields.cardProperties = [
        {
            id: 'property_id_1',
            name: 'Owner',
            type: 'select',
            options: [
                {
                    color: 'propColorDefault',
                    id: 'property_value_id_1',
                    value: 'Jean-Luc Picard',
                },
                {
                    color: 'propColorDefault',
                    id: 'property_value_id_2',
                    value: 'William Riker',
                },
                {
                    color: 'propColorDefault',
                    id: 'property_value_id_3',
                    value: 'Deanna Troi',
                },
            ],
        },
        {
            id: 'property_id_2',
            name: 'MockStatus',
            type: 'number',
            options: [],
        },
    ]

    const view = TestBlockFactory.createBoardView(board)
    view.fields.sortOptions = []
    view.fields.groupById = undefined
    view.fields.hiddenOptionIds = []

    const card = TestBlockFactory.createCard(board)
    card.fields.properties.property_id_1 = 'property_value_id_1'
    card.fields.properties.property_id_2 = 'property_value_id_2'

    const cardTemplate = TestBlockFactory.createCard(board)
    cardTemplate.fields.isTemplate = true

    function renderComponent() {
        const component = wrapIntl((
            <CardDetailProperties
                board={board!}
                card={card}
                cards={[card]}
                contents={[]}
                comments={[]}
                activeView={view}
                views={[view]}
                readonly={false}
            />
        ))

        const result = render(component)
        return result
    }

    test('should match snapshot', async () => {
        const container = renderComponent().container

        expect(container).toMatchSnapshot()
    })

    test('rename select property and confirm on dialog should rename property', async () => {
        const result = renderComponent()

        onPropertyRenameOpenConfirmationDialog(result.container)

        const confirmButton = result.getByTitle('Change Property')
        expect(confirmButton).toBeDefined()

        userEvent.click(confirmButton!)

        // should be called once on confirming renaming the property
        expect(FetchMock.fn).toBeCalledTimes(1)

        // Verify API call was made with renamed property
        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
        // @ts-ignore
        const lastAPICallPayload = JSON.parse(FetchMock.fn.mock.calls[0][1].body)
        expect(lastAPICallPayload[0].fields.cardProperties[0].name).toBe('Owner - Renamed')
        expect(lastAPICallPayload[0].fields.cardProperties[0].options.length).toBe(3)
        expect(lastAPICallPayload[0].fields.cardProperties[0].options[0].value).toBe('Jean-Luc Picard')
        expect(lastAPICallPayload[0].fields.cardProperties[0].options[1].value).toBe('William Riker')
        expect(lastAPICallPayload[0].fields.cardProperties[0].options[2].value).toBe('Deanna Troi')
    })

    test('cancel btn in TypeorNameChange dialog should do nothing', () => {
        const result = renderComponent()
        const container = result.container
        onPropertyRenameOpenConfirmationDialog(container)

        const cancelButton = result.getByTitle('Cancel')
        expect(cancelButton).toBeDefined()

        userEvent.click(cancelButton!)

        expect(container).toMatchSnapshot()
    })

    test('confirmation on delete dialog should delete the property', () => {
        const result = renderComponent()
        const container = result.container

        openDeleteConfirmationDialog(container)

        const confirmButton = result.getByTitle('Delete')
        expect(confirmButton).toBeDefined()

        //click delete button
        userEvent.click(confirmButton!)

        // should be called once on confirming delete
        expect(FetchMock.fn).toBeCalledTimes(1)

        // Verify API call was made and deleted property does not existing
        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
        // @ts-ignore
        const lastAPICallPayload = JSON.parse(FetchMock.fn.mock.calls[0][1].body)
        const deletedPropInPayload = lastAPICallPayload[0].fields.cardProperties.filter(
            (prop:IPropertyTemplate) => prop.id === board.fields.cardProperties[0].id)

        expect(deletedPropInPayload.length).toBe(0)
    })

    test('cancel on delete dialog should do nothing', () => {
        const result = renderComponent()
        const container = result.container

        openDeleteConfirmationDialog(container)

        const cancelButton = result.getByTitle('Cancel')
        expect(cancelButton).toBeDefined()

        userEvent.click(cancelButton!)
        expect(container).toMatchSnapshot()
    })

    function openDeleteConfirmationDialog(container:HTMLElement) {
        const propertyLabel = container.querySelector('.MenuWrapper')
        expect(propertyLabel).toBeDefined()
        fireEvent.click(propertyLabel!)

        const deleteOption = container.querySelector('.MenuOption.TextOption')
        expect(propertyLabel).toBeDefined()
        userEvent.click(deleteOption!)

        const confirmDialog = container.querySelector('.dialog.confirmation-dialog-box')
        expect(confirmDialog).toBeDefined()
    }

    function onPropertyRenameOpenConfirmationDialog(container:HTMLElement) {
        const propertyLabel = container.querySelector('.MenuWrapper')
        expect(propertyLabel).toBeDefined()
        fireEvent.click(propertyLabel!)

        // write new name in the name text box
        const propertyNameInput = container.querySelector('.PropertyMenu.menu-textbox')
        expect(propertyNameInput).toBeDefined()
        userEvent.type(propertyNameInput!, 'Owner - Renamed{enter}')
        fireEvent.click(propertyLabel!)

        const confirmDialog = container.querySelector('.dialog.confirmation-dialog-box')
        expect(confirmDialog).toBeDefined()
    }
})

