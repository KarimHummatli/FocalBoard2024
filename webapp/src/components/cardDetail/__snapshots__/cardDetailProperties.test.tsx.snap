// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/cardDetail/CardDetailProperties cancel btn in TypeorNameChange dialog should do nothing 1`] = `
<div>
  <div
    class="octo-propertylist CardDetailProperties"
  >
    <div
      class="octo-propertyrow"
    >
      <div
        aria-label="menuwrapper"
        class="MenuWrapper"
        role="button"
      >
        <div
          class="octo-propertyname"
        >
          <button
            class="Button"
            type="button"
          >
            <span>
              Owner
            </span>
          </button>
        </div>
      </div>
      <div
        class="octo-propertyvalue"
        data-testid="select-non-editable"
        tabindex="0"
      >
        <span
          class="Label propColorDefault "
        >
          <span
            class="Label-text"
          >
            Jean-Luc Picard
          </span>
        </span>
      </div>
    </div>
    <div
      class="octo-propertyrow"
    >
      <div
        aria-label="menuwrapper"
        class="MenuWrapper"
        role="button"
      >
        <div
          class="octo-propertyname"
        >
          <button
            class="Button"
            type="button"
          >
            <span>
              MockStatus
            </span>
          </button>
        </div>
      </div>
      <input
        class="Editable error octo-propertyvalue"
        placeholder="Empty"
        spellcheck="false"
        style="width: 5px;"
        title="property_value_id_2"
        value="property_value_id_2"
      />
    </div>
    <div
      class="octo-propertyname add-property"
    >
      <button
        class="Button"
        type="button"
      >
        <span>
          + Add a property
        </span>
      </button>
    </div>
  </div>
</div>
`;

exports[`components/cardDetail/CardDetailProperties cancel on delete dialog should do nothing 1`] = `
<div>
  <div
    class="octo-propertylist CardDetailProperties"
  >
    <div
      class="octo-propertyrow"
    >
      <div
        aria-label="menuwrapper"
        class="MenuWrapper"
        role="button"
      >
        <div
          class="octo-propertyname"
        >
          <button
            class="Button"
            type="button"
          >
            <span>
              Owner
            </span>
          </button>
        </div>
      </div>
      <div
        class="octo-propertyvalue"
        data-testid="select-non-editable"
        tabindex="0"
      >
        <span
          class="Label propColorDefault "
        >
          <span
            class="Label-text"
          >
            Jean-Luc Picard
          </span>
        </span>
      </div>
    </div>
    <div
      class="octo-propertyrow"
    >
      <div
        aria-label="menuwrapper"
        class="MenuWrapper"
        role="button"
      >
        <div
          class="octo-propertyname"
        >
          <button
            class="Button"
            type="button"
          >
            <span>
              MockStatus
            </span>
          </button>
        </div>
      </div>
      <input
        class="Editable error octo-propertyvalue"
        placeholder="Empty"
        spellcheck="false"
        style="width: 5px;"
        title="property_value_id_2"
        value="property_value_id_2"
      />
    </div>
    <div
      class="octo-propertyname add-property"
    >
      <button
        class="Button"
        type="button"
      >
        <span>
          + Add a property
        </span>
      </button>
    </div>
  </div>
</div>
`;

exports[`components/cardDetail/CardDetailProperties should match snapshot 1`] = `
<div>
  <div
    class="octo-propertylist CardDetailProperties"
  >
    <div
      class="octo-propertyrow"
    >
      <div
        aria-label="menuwrapper"
        class="MenuWrapper"
        role="button"
      >
        <div
          class="octo-propertyname"
        >
          <button
            class="Button"
            type="button"
          >
            <span>
              Owner
            </span>
          </button>
        </div>
      </div>
      <div
        class="octo-propertyvalue"
        data-testid="select-non-editable"
        tabindex="0"
      >
        <span
          class="Label propColorDefault "
        >
          <span
            class="Label-text"
          >
            Jean-Luc Picard
          </span>
        </span>
      </div>
    </div>
    <div
      class="octo-propertyrow"
    >
      <div
        aria-label="menuwrapper"
        class="MenuWrapper"
        role="button"
      >
        <div
          class="octo-propertyname"
        >
          <button
            class="Button"
            type="button"
          >
            <span>
              MockStatus
            </span>
          </button>
        </div>
      </div>
      <input
        class="Editable error octo-propertyvalue"
        placeholder="Empty"
        spellcheck="false"
        style="width: 5px;"
        title="property_value_id_2"
        value="property_value_id_2"
      />
    </div>
    <div
      class="octo-propertyname add-property"
    >
      <button
        class="Button"
        type="button"
      >
        <span>
          + Add a property
        </span>
      </button>
    </div>
  </div>
</div>
`;
